clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -277,6 +277,7 @@ else if (expireMs <= now) {
     public synchronized void failedUpdate(long now, AuthenticationException authenticationException) {
         this.lastRefreshMs = now;
         this.authenticationException = authenticationException;
+        this.version += 1;
```

Comment:
Have you verified that this works for all uses of version at the moment? It seems like a minor change in behavior since we use `failedUpdate` in `DefaultMetadataUpdater` if there are no nodes in the response. Currently this would cause uses such as `ConsumerNetworkClient.awaitMetadataUpdate()` to retry before returning.

Reply:
I don't think retrying is what we want, though.  If you get an authentication exception from any broker, your auth is bad and you should fail.  It's not a retryable exception, conceptually or in terms of code

Reply:
The case I am referring to is in `DefaultMetadataupdater.handleCompletedMetadataResponse`, which is different from the authentication failure path. We have a check to ensure that the metadata response contains at least one broker. With this change, we will update the metadata version even in the case that it is empty, which will cause methods like `ConsumerNetworkClient.awaitMetadataUpdate()` to return earlier than they currently do.

Reply:
Hmm.  Maybe I'm misinterpreting, but are you suggesting that `ConsumerNetworkClient#awaitMetadataUpdate()` should wait forever (or until the timeout hits) when there is an authentication error fetching the metadata?  That doesn't seem right.

Reply:
I am definitely not suggesting that. I am not talking about authentication failures at all. I am referring to the call to `Metadata.failedUpdate` that is in `DefaultMetadataupdater.handleCompletedMetadataResponse`. With this change, the version will be incremented following an "empty" metadata update. This will cause us to exit the loop in `ConsumerNetworkClient#awaitMetadataUpdate()` even though we have not received a valid update. 

My point more generally is that we are changing the meaning of the version field inside `Metadata`. It seems we were intentionally using this before to indicate when we had received a valid new version of the metadata. But now we bump it even when there's a failure which means we can no longer use it to tell when we've seen a successful update. As far as I can tell, the impact may be minor, but we should consider all of the usages to be sure of it.